qiskit/transpiler/passes/optimization/light_cone.py

```diff
@@ -0,0 +1,141 @@
+from qiskit.circuit.commutation_library import standard_gates_commutations
+from qiskit.circuit.controlflow import CONTROL_FLOW_OP_NAMES
+from qiskit.circuit.instruction import Instruction
+from qiskit.circuit.library.generalized_gates.pauli import PauliGate
+from qiskit.circuit.library.standard_gates.p import PhaseGate
+from qiskit.circuit.library.standard_gates.rx import RXGate
+from qiskit.circuit.library.standard_gates.rz import RZGate
+from qiskit.circuit.library.standard_gates.u1 import U1Gate
+from qiskit.circuit.quantumregister import QuantumRegister, Qubit
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.dagcircuit import DAGCircuit, DAGInNode, DAGNode, DAGOutNode
+from qiskit.quantum_info import Pauli
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.passes.optimization.commutation_analysis import \
+    CommutationAnalysis
+from qiskit.transpiler.passmanager import PassManager
+
+commutator = CommutationChecker(standard_gates_commutations)
+
+
+class LightCone(TransformationPass):
+    """Remove the gates that do not affect the outcome of a measurement on a circuit.
+
+    Pass for computing the light-cone of an observable or measurement. The Pass can
+    handle either a Pauli operator one would like to measure or a measurement on a set
+    of qubits.
+    """
+
+    def __init__(self, observable: Pauli | None = None):
+        """
+        LightCone initializer.
+
+        Args:
+            observable: If None the lightcone will be computed for the set
+            of measurements in the circuit.
+            If a Pauli operator is specified, the lightcone will correspond to
+            the reduced circuit with the
+            same expectation value for the observable.
+        """
+        super().__init__()
+        self.observable = observable
+
+    @staticmethod
+    def _find_measurement_qubits(dag: DAGCircuit):
+        """For now this method assumes that all the measurements in the circuit are
+        final measurements."""
+        qubits_measured = set()
+        for node in dag.topological_nodes():
+            # print(getattr(node, "name", False))
+            if getattr(node, "name", False) == "measure":
+                qubits_measured |= {x for x in node.qargs}
+        print(qubits_measured)
+        return qubits_measured
+
+    def _get_initial_lightcone(
+        self, dag: DAGCircuit
+    ) -> tuple[set[int], tuple[Instruction, list[int]]]:
+        """Returns the initial lightcone.
+        If obervable is None, the lightcone is a different Z operator in each one of the
+        qubits with a measurement. If a Pauli observable is defined, that will be
+        the first gate in the circuit.
+        """
+        if self.observable is None:
+            qubits_measured = self._find_measurement_qubits(dag)
+            light_cone = qubits_measured
+            light_cone_ops = [(PauliGate("Z"), [qubit_index]) for qubit_index in qubits_measured]
```

Comment:
It's likely better to use `ZGate` here so we can access the pre-computed commutations of the commutation checker. I don't think these are defined for the PauliGate.
```suggestion
            light_cone_ops = [(ZGate(), [qubit_index]) for qubit_index in qubits_measured]
```